Handle "nested" RESTful APIs

jimdillon opened the pull request:
- Add opt.pathToPrepend to syncOpts to handle nested RESTful APIs
- Fix tests for path
- Update README.md

jimdillon committed:
Add opt.pathToPrepend to handle nested RESTful APIs

evanshortiss commented:
Hey @jimdillon, can you briefly explain the need for this? Would structuring your app.use differently and using the _noTrim_ not suffice?

evanshortiss commented:
@jimdillon I read the repo name wrong and am on mobile, hence getting it mixed up with fh-mbaas-proxy. I see why this is useful since it enables you to configure more complex routing structures. It's likely I'll merge it, but can guarantee a timeline before next week so I'd say use you fork for now it shouldn't diverge much.

jimdillon commented:
No worries @evanshortiss.  Enjoy your vacation, this can definitely wait.

jim
On Jul 1, 2016 8:00 AM, "Evan Shortiss" notifications@github.com wrote:

> @jimdillon https://github.com/jimdillon I read the repo name wrong and
> am on mobile, hence getting it mixed up with fh-mbaas-proxy. I see why this
> is useful since it enables you to configure more complex routing
> structures. It's likely I'll merge it, but can guarantee a timeline before
> next week so I'd say use you fork for now it shouldn't diverge much.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> https://github.com/feedhenry-staff/fh-rest-sync-proxy/pull/1#issuecomment-229929229,
> or mute the thread
> https://github.com/notifications/unsubscribe/AMnRtmFkIjxdRt5V83Apglb2fnrMq4A_ks5qRQF1gaJpZM4JCkJs
> .

evanshortiss commented on package.json:
```diff
@@ -12,7 +12,9 @@
-  "author": "Evan Shortiss",
+  "author": {
+    "name": "Evan Shortiss"
+  },
```
Can you add a new field here "contributors" and add yourself? Here's the docs: https://docs.npmjs.com/files/package.json#people-fields-author-contributors

evanshortiss commented on lib/handlers.js:
```diff
@@ -72,7 +76,7 @@ module.exports = function (dataset, opts) {
       method: 'GET',
       params: params,
-      path: '/' + dataset
+      path: pathMod.normalize('/' + opts.pathToPrepend + '/' + dataset)
```
I recently learned the _path_ module will change forward slash to backslash when used on windows. Unlikely to be an issue for us since we know 99% of the time people use Ubuntu or Mac. We could instead use something like this to safely join the url path segments: https://www.npmjs.com/package/url-join

I've not tested it, but looks like it could work.

evanshortiss commented on lib/handlers.js:
```diff
@@ -96,7 +100,7 @@ module.exports = function (dataset, opts) {
       guid: opts.guid,
       method: 'GET',
-      path: '/' + dataset + '/' + id.toString()
+      path: pathMod.normalize('/' + opts.pathToPrepend + '/' + dataset + '/' + id.toString())
```
Probably should update the tests to verify this path is constructed correctly based on the inclusion or omission of _pathToPrepend_.

jimdillon replied:
@evanshortiss Good points, it'll be Monday before I have any changes pushed.

jimdillon commented:
@evanshortiss, Added contributors property in package.json as requested.

jimdillon committed:
use url-join and improve testing
Change title in README.md to fh-rest-sync-proxy
update package.json to include url-join
Change coverage percentages, statements 98, lines 98

jimdillon commented:
@evanshortiss Ready for review.

evanshortiss commented on package.json:
```diff
@@ -6,16 +6,22 @@
+  "author": {
+    "name": "Evan Shortiss"
+  },
+  "contributors": {
```
The current Object is fine here, just needs to be wrapped in an Array 👍 

```
[
  {
    "name": "Jim Dillon"
  }
]
```

evanshortiss commented on test/handlers.test.js:
```diff
@@ -128,7 +214,26 @@ describe(__filename, function () {
+    });
+  });
+
+  describe('#handleDelete', function () {
```
Thanks for updating these tests. I think the new structure is unusual though. We have multiple _describe_ calls for the same functions. Instead, could you use multiple _it_ calls?

For example:

``` js
describe('#handleDelete', function () {
  it('should work without a nested path', function (done) { /*etc. */});

  it('should work with a nested path', function (done) { /*etc. */})
});
```

I also think you only need one new test case for the path handling if you abstract it into a small module rather than making a new test for each function? It would also provide a cleaner API since you could just omit _pathToPrepend_ and not need to pass an empty string.

Example:

``` js

module.exports = function buildRoute (opts, dataset) {
  return (opts.pathToPrepend) ? (opts.pathToPrepend + dataset) :  dataset;
};

```

evanshortiss replied:
That example for opts.pathToPrepend is invalid, it needs to do the urlJoin stuff, but hope it makes sense?

evanshortiss commented on README.md:
```diff
@@ -1,4 +1,4 @@
-# fh-sync-mbaas-proxy
+# fh-rest-sync-proxy
```
Nice catch! 😂

jimdillon commented:
@evanshortiss Care to take another gander?

evanshortiss commented on lib/build-route.js:
```diff
@@ -0,0 +1,14 @@
+
+module.exports = function buildRoute(opts, dataset) {
+  if(opts && opts.pathToPrepend) {
+    console.log('PATH ', urlJoin(opts.pathToPrepend, dataset));
```
Can you delete those logs?

evanshortiss commented on lib/handlers.js:
```diff
@@ -2,16 +2,19 @@
 
   log.debug(
-    'creating MBaaS sync proxy for dataset "%s" with guid "%s"',
+    'creating MBaaS sync proxy for dataset "%s" with guid "%s" prepend "%s"',
```
Would be nice to change this to read:

```
creating MBaaS sync proxy for dataset "orders" with guid "GUID" to endpoint "root/path"
```

evanshortiss replied:
The reason being, I think prepend doesn't read well and could cause a little confusion without full context

evanshortiss commented on lib/handlers.js:
```diff
@@ -27,9 +30,10 @@ module.exports = function (dataset, opts) {
         var e = new VError(
           err,
-          'failed to perform "%s" for "%s"',
+          'failed to perform "%s" for "%s" prepend "%s"',
```
Similar here.

evanshortiss commented on lib/handlers.js:
```diff
@@ -40,10 +44,11 @@ module.exports = function (dataset, opts) {
       } else {
         log.debug(
-          'sync call "%s" for "%s" was successful to "%s"',
+          'sync call "%s" for "%s" was successful to "%s" prepend "%s"',
```
And here 😄

jimdillon committed:
Remove console.log (s).  Improve error messages

jimdillon commented:
@evanshortiss one more for the win?
